### PR TITLE
8274274: Update JUnit to version 5.8.1

### PR DESCRIPTION
**Pull request by hjohn**
I've added JUnit 5 as a test dependency and made sure that the JUnit 4 tests still work.  Also added a single JUnit 5 tests, and confirmed it works.

I've updated the Eclipse project file for the base module only.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8274274](https://bugs.openjdk.java.net/browse/JDK-8274274): Update JUnit to version 5.8.1

### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Michael Strauß](https://openjdk.java.net/census#mstrauss) (@mstr2 - Author)
 * [Johan Vos](https://openjdk.java.net/census#jvos) (@johanvos - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/633/head:pull/633` \
`$ git checkout pull/633`

Update a local copy of the PR: \
`$ git checkout pull/633` \
`$ git pull https://git.openjdk.java.net/jfx pull/633/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 633`

View PR using the GUI difftool: \
`$ git pr show -t 633`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/633.diff">https://git.openjdk.java.net/jfx/pull/633.diff</a>

</details>

### CODE REVIEW

File: gradle/verification-metadata.xml
Comment (kevinrushforth) [inline gradle/verification-metadata.xml]: Why is junit-4.13 needed?

File: gradle/verification-metadata.xml
Comment (kevinrushforth) [inline gradle/verification-metadata.xml]: Please revert this. `apiguardian` is not an approved third-party library.

File: gradle/verification-metadata.xml
Comment (kevinrushforth) [inline gradle/verification-metadata.xml:273]: Is `hamcrest` actually required in order to support JUnit 5? If not, this should be reverted. If so, that's another third-party dependency we would need to get approval for.

File: gradle/verification-metadata.xml
Comment (kevinrushforth) [inline gradle/verification-metadata.xml:345]: Is `opentest4j` actually required in order to support JUnit 5? If not, this should be reverted. If so, that's another third-party dependency we would need to get approval for.

**Comment by hjohn** [comment]
Sorry for the force push, forgot I already submitted it as a PR.  I only removed a few extraneous dependencies that got added in the `verification-metadata.xml` (a few extra got added from different versions of JUnit 5 that weren't needed for the final version I'm using).

File: gradle/verification-metadata.xml
Comment (kevinrushforth) [inline gradle/verification-metadata.xml:121]: Why the dependency on junit 4.13.2?

File: gradle/verification-metadata.xml
Comment (kevinrushforth) [inline gradle/verification-metadata.xml:247]: Is `apiguardian` actually required in order to support JUnit 5? If not, this should be reverted. If so, that's another third-party dependency we would need to get approval for.

File: modules/javafx.base/src/test/java/test/JUnit5Test.java
Comment (kevinrushforth) [inline modules/javafx.base/src/test/java/test/JUnit5Test.java:1]: This needs a copyright header block.

**Review by kevinrushforth — CHANGES_REQUESTED** [review]
As mentioned in JBS, any new third-party libraries require prior third-party license approval. And we will need to work with you on sponsoring this (as you can't contribute any third-party code under the OCA).

Speaking of which, there are more libraries added to `gradle-verification.xml` than I would expect. Each one will need third-party approval, if they are required. Since this is for internal use (build / test only) and not something we redistribute, that makes it easier, although it still depends on the license for each piece.

Finally, we have some closed tests that we need to ensure aren't impacted by this, so that's one more area we need to coordinate.

We can take a look, but it won't be right away. Can you also start a thread on openjfx-dev? I'd like to gauge the level of interest in enabling JUnit 5 for new tests.

**Comment by kevinrushforth** [comment]
> Sorry for the force push, forgot I already submitted it as a PR.

That's OK. I deleted the couple of comments I had added relating to blocks that disappeared.

File: gradle/verification-metadata.xml
Comment (hjohn) [reply gradle/verification-metadata.xml:247]: It's an annotation used within JUnit code, it might be possible to block this one, I will see how to do this in Gradle syntax.

This is the whole transitive tree it pulls in:
image

**Comment by kevinrushforth** [comment]
Btw, I think this should be moved to a Draft PR until the third-party issues are resolved (including any needed approvals).

File: gradle/verification-metadata.xml
Comment (hjohn) [reply gradle/verification-metadata.xml:273]: I will try block this and get back to you.

File: gradle/verification-metadata.xml
Comment (hjohn) [reply gradle/verification-metadata.xml:345]: Will try block

File: gradle/verification-metadata.xml
Comment (hjohn) [reply gradle/verification-metadata.xml:121]: The vintage engine pulls this in, so I think we may be forced to upgrade from 4.8.2 JUnit 4 to 4.13.2 JUnit 4 as well, although the current setup seems to work (I remember some Rules being deprecated in later JUnit 4, specifically rules for catching exceptions which could require rewriting those tests to use `assertThrows`).

File: gradle/verification-metadata.xml
Comment (hjohn) [reply gradle/verification-metadata.xml:324]: I couldn't block opentest4j.  Apparently this is a project that harmonizes the exceptions thrown by various test frameworks (not just JUnit 5) and JUnit 5 depends on it for the exceptions it uses to signal test failures and other error conditions.  It is a fairly small project (6 classes only) licensed under the Apache 2.0 license: https://github.com/ota4j-team/opentest4j

File: modules/javafx.base/src/test/java/test/JUnit5Test.java
Comment (hjohn) [reply modules/javafx.base/src/test/java/test/JUnit5Test.java:26]: Added

File: gradle/verification-metadata.xml
Comment (hjohn) [reply gradle/verification-metadata.xml]: Blocked this one without ill effects, tests still run.

File: gradle/verification-metadata.xml
Comment (hjohn) [reply gradle/verification-metadata.xml]: Hamcrest could be blocked, tests still run.

**Comment by kevinrushforth** [comment]
The gradle verification task failed (see the test results from GitHub Actiona). It looks like it still needs the hamcrest jars. Go ahead and add them back in (and remove the exclusion for them).

**Comment by hjohn** [comment]
> The gradle verification task failed (see the test results from GitHub Actiona). It looks like it still needs the hamcrest jars. Go ahead and add them back in (and remove the exclusion for them).

Yeah, I noticed, I couldn't get rid of them.  Apparently JUnit 4 itself is depending on them now (starting from 4.10 orso) and the Vintage engine requires JUnit 4.12+, so it is actually overriding JUnit 4.8.2 to 4.13.2 and pulls in the Hamcrest jars again.

I've tried forcing JUnit to 4.8.2 and even though that works, it leads to a spam stacktrace in the logs for each test executed because it can't find something via reflection.  I then tried to hide this log message, but have been unable to find any way of getting that message blocked by configuring java.util.logging that works in combination with Gradle and its daemon.

So... I've sort of think this is not a fight we can't win easily and have reincluded the Hamcrest jars.  I also removed the JUnit 4 dependency as the Vintage engine will pull a newer one in any way.

I think I should upgrade these lines as well: `testImplementation group: "junit", name: "junit", version: "4.8.2"` in each of the modules to make it clear 4.13.2 is used and not 4.8.2.

**Comment by kevinrushforth** [comment]
All tasks are now completed. You can move this PR back to ready and the review can proceed.

File: gradle/verification-metadata.xml
Comment (hjohn) [reply gradle/verification-metadata.xml:121]: Assuming this is okay now as approvals have been received.

Comment (kevinrushforth) [reply gradle/verification-metadata.xml:121]: Yes, this is fine.

File: gradle/verification-metadata.xml
Comment (kevinrushforth) [reply gradle/verification-metadata.xml:265]: As it turns out, `hamcrest` does not need to be blocked, so you can restore it if it is easier. I'll leave it up to you.

File: gradle/verification-metadata.xml
Comment (kevinrushforth) [reply gradle/verification-metadata.xml:337]: Since we only use JUnit for build / test, and don't directly depend on `opentest4j`, this is not a problem.

File: build.gradle
Comment (kevinrushforth) [inline build.gradle:1953]: As it turns out, `apiguardian` no longer needs to be blocked, so you can restore it if it is easier.

**Review by kevinrushforth — COMMENTED** [review]
I left a few comments on the dependencies. Will review / test the PR later.

One comment about adding new JUnit 5 tests and migrating existing tests. I think there could be value in organizing the tests such that all of the JUnit 5 tests are grouped, rather than mixing tests in the same directory such that some use JUnit 5 and others use JUnit 4. What do you (or others) think? We could either do this with a new JUnit 5 source set in each project, or by using a package naming convention for JUnit 5 tests like we do for robot tests. Maybe `test5.some.pkg`. This needs more thought.

**Comment by mstr2** [comment]
> I left a few comments on the dependencies. Will review / test the PR later.
> 
> One comment about adding new JUnit 5 tests and migrating existing tests. I think there could be value in organizing the tests such that all of the JUnit 5 tests are grouped, rather than mixing tests in the same directory such that some use JUnit 5 and others use JUnit 4. What do you (or others) think? We could either do this with a new JUnit 5 source set in each project, or by using a package naming convention for JUnit 5 tests like we do for robot tests. Maybe `test5.some.pkg`. This needs more thought.

I don't like the idea of separating source sets or packages just based on the version of our unit test framework. In my opinion, sources should be organized by function, and not based on a technicality. A very common case we'll encounter will be adding a new unit test to an existing test class. What's our guidance in this case? If we continue to use JUnit4 for new unit tests, that means we'll effectively persist this version of the framework forever in the test sources. Another option would be to migrate the entire test class once it's updated with new code. However, that would be a significant amount of work required from the author of a new unit test. A third option would be to actually migrate all unit test classes as a series of refactoring PRs.

**Comment by hjohn** [comment]
> One comment about adding new JUnit 5 tests and migrating existing tests. I think there could be value in organizing the tests such that all of the JUnit 5 tests are grouped, rather than mixing tests in the same directory such that some use JUnit 5 and others use JUnit 4. What do you (or others) think? We could either do this with a new JUnit 5 source set in each project, or by using a package naming convention for JUnit 5 tests like we do for robot tests. Maybe `test5.some.pkg`. This needs more thought.

I don't think this is really needed, since the introduction of JUnit 5 I've written numerous tests in projects with the bulk of their test classes still using JUnit 4. In all cases, the new tests were just mixed with the old tests. The mixing of the tests has so far caused no confusion, and in these projects the teams involved also didn't feel any need to prioritize migrating all tests to JUnit 5. 

Conversion to JUnit 5 usually happens on a test by test basis when other changes are made to the test -- existing code is rewritten as needed, usually limited to introducing `assertThrows` to replace `@Test(expected = Throwable.class)` and updating all static asserts to use the `jupiter` package hierarchy instead of the `org.junit` one (just import changes).  Usually this is done in a separate commit (a "chore" commit) with another commit containing the actual change and new test cases.

**Comment by kevinrushforth** [comment]
These are good points regarding test migration. I withdraw the suggestion of test segregation.

I would not want to see a wholesale migration of all tests to JUnit5, but I do like the idea of migrating on a test-by-test basis if new tests are being added to an existing test class where using JUnit 5 would be a benefit.

File: build.gradle
Comment (hjohn) [reply build.gradle:1953]: Okay, I'll remove the exclusion on apiguardian then as it would just be confusing why it was excluded in the first place.  Hamcrest was already included as JUnit 5 wouldn't work without it.

**Comment by hjohn** [comment]
> As mentioned in JBS, any new third-party libraries require prior third-party license approval. And we will need to work with you on sponsoring this (as you can't contribute any third-party code under the OCA).
> 
> Speaking of which, there are more libraries added to `gradle-verification.xml` than I would expect. Each one will need third-party approval, if they are required. Since this is for internal use (build / test only) and not something we redistribute, that makes it easier, although it still depends on the license for each piece.
> 
> Finally, we have some closed tests that we need to ensure aren't impacted by this, so that's one more area we need to coordinate.
> 
> We can take a look, but it won't be right away. Can you also start a thread on openjfx-dev? I'd like to gauge the level of interest in enabling JUnit 5 for new tests.

Github is showing this as a "requested change" but I can't resolve this.  @kevinrushforth anything you can do on your end?

All the issues in this comment were addresses AFAIK.

**Review by kevinrushforth — COMMENTED** [review]
I was able to do a build / test on my local build machine using locally cached jar files. In order for this to work, I needed to add an explicit list of dependencies in `build.gradle`. See comments inline.

Once you make the requested updates, I'll do a CI test build.

**Comment by kevinrushforth** [comment]
I also was able to make the (minor) changes needed to our closed build to use the `4.13.2` jar files.

As a note to other committers: Given that this is a third-party update, and that I need to integrate a coordinated closed change, I will sponsor this once it is ready. Please don't sponsor it even if it appears to be waiting for a while.

File: modules/javafx.base/src/test/java/test/JUnit5Test.java
Comment (hjohn) [reply modules/javafx.base/src/test/java/test/JUnit5Test.java:36]: Added

File: build.gradle
Comment (hjohn) [reply build.gradle]: Removed this and in the other location.

File: build.gradle
Comment (hjohn) [reply build.gradle]: Fixed

File: build.gradle
Comment (hjohn) [reply build.gradle:1949]: I've added the dependencies you listed here.

File: modules/javafx.base/src/test/java/test/JUnit5Test.java
Comment (kevinrushforth) [inline modules/javafx.base/src/test/java/test/JUnit5Test.java:28]: For JUnit 5, shouldn't this be `org.junit.jupiter.api.Assertions.assertNotNull`?

**Review by kevinrushforth — COMMENTED** [review]
Looks good with one comment inline.

**Comment by kevinrushforth** [comment]
I was looking for places where JUnit 4.8.2 was still being referenced and found two more places that should be changed:

1. [`buildSrc/build.gradle`](https://github.com/openjdk/jfx/blob/d06d657007c3af86ba2ff908b546cae18e775bfb/buildSrc/build.gradle#L77) depends on JUnit 4.8.2 -- this should be changed to JUnit 4.13.2.
2. Once the above is fixed, you should be able to remove the JUnit 4.8.2 artifacts from [`gradle/verification-metadata.xml`](https://github.com/openjdk/jfx/blob/d06d657007c3af86ba2ff908b546cae18e775bfb/gradle/verification-metadata.xml#L122)

File: modules/javafx.base/src/test/java/test/JUnit5Test.java
Comment (hjohn) [reply modules/javafx.base/src/test/java/test/JUnit5Test.java:28]: Sorry, I totally missed that.  Although you can mix these without ill effect, we should move towards the `jupiter` annotations if we ever want to get to the point where JUnit 4 can be phased out.

**Review by kevinrushforth — APPROVED** [review]
Looks good. I did a CI build with the latest changes.

**Comment by kevinrushforth** [comment]
@johanvos @tiainen Can one of you also review this to ensure it causes no problems with your builds?

**Review by mstr2 — APPROVED** [review]
The changes look good to me. The tests work as expected.

**Comment by johanvos** [comment]
This looks good. I noticed that the fix for 
8276142: Update gradle to version 7.3
was not in the branch for this PR, but merging it manually  (which will be done when it is merged here) worked and did not cause issues.
I'm running a few more tests, and will approve when they are all green.

**Review by johanvos — APPROVED** [review]

**Pull request closed without merging**
